aleo-programs/staking_lite/src/main.leo

```diff
@@ -0,0 +1,376 @@
+    * The height at which the current withdrawal batch will be done unbonding
+    * if not present or == 0u32, a new batch can begin unbonding
+  */
+  mapping current_batch_height: u8 => u32;
+
+  struct withdrawal_state {
+    microcredits: u64;
+    claim_block: u32;
+  }
+
+  // address -> pending withdrawal for the delegator with this address
+  mapping withdrawals: address => withdrawal_state;
+
+  transition initialize(commission_rate: u128, validator_address: address) {
+    assert_eq(self.caller, ADMIN);
+    assert(commission_rate < PRECISION_UNSIGNED);
+
+    return then finalize(commission_rate, validator_address);
+  }
+
+  finalize initialize(commission_rate: u128, validator_address: address) {
+    commission_percent.set(0u8, commission_rate);
+    validator.set(0u8, validator_address);
+    state.set(0u8, 0u8);
+    total_shares.set(0u8, 0u64);
+    total_balance.set(0u8, 0u64);
+    pending_withdrawal.set(0u8, 0u64);
+    current_batch_height.set(0u8, 0u32);
+  }
+
+  transition initial_deposit(
+    input_record: credits.aleo/credits,
+    microcredits: u64,
+    validator_address: address
+  ) -> credits.aleo/credits {
+    assert_eq(self.caller, ADMIN);
+    // Must be a credits record because credits.aleo uses self.caller for transfers
+    let updated_record: credits.aleo/credits = credits.aleo/transfer_private_to_public(input_record, CORE_PROTOCOL, microcredits);
+    credits.aleo/bond_public(validator_address, microcredits);
+
+    return (updated_record) then finalize(microcredits);
+  }
+
+  finalize initial_deposit(microcredits: u64) {
+    let balance: u64 = total_balance.get_or_use(0u8, 0u64);
+    let shares: u64 = total_shares.get_or_use(0u8, 0u64);
+    assert_eq(balance, 0u64);
+    assert_eq(shares, 0u64);
+
+    total_balance.set(0u8, microcredits);
+    total_shares.set(0u8, microcredits);
+    delegator_shares.set(ADMIN, microcredits);
+  }
+
+  transition set_commission_percent(commission_rate: u128) {
+    assert_eq(self.caller, ADMIN);
+    assert(commission_rate < PRECISION_UNSIGNED);
+
+    return then finalize(commission_rate);
+  }
+
+  finalize set_commission_percent(commission_rate: u128) {
+    let current_state: u8 = state.get(0u8);
+    assert_eq(current_state, 0u8);
+
+    // Make sure all commission is claimed before changing the rate
+    let balance: u64 = total_balance.get(0u8);
+    let bonded: u64 = 0u64; // credits.aleo/bonded.get(CORE_PROTOCOL);
+    assert_eq(balance, bonded);
+
+    commission_percent.set(0u8, commission_rate);
+  }
+
+  transition unbond_all(pool_balance: u64) {
+    assert_eq(self.caller, ADMIN);
+    credits.aleo/unbond_public(pool_balance);
+
+    return then finalize();
+  }
+
+  finalize unbond_all() {
+    // Make sure all commission is claimed before unbonding
+    let balance: u64 = total_balance.get(0u8);
+    let bonded: u64 = 0u64; // credits.aleo/bonded.get(CORE_PROTOCOL);
+    assert_eq(balance, bonded);
+
+    state.set(0u8, 1u8);
```

Comment:
Do we need another assert that we're not already in a different state?

Reply:
I'm a little uncertain here. Before the dev call today, unbond_all should have been callable multiple times for saftey. There was a case where unbond_all may be called, but rewards (>10 credits) are disbursed and not all the bonded credits are actually unbonded, so an additional unbond call would have to be made to prevent the protocol from getting stuck.

After the dev call today, this case may no longer happen since the bond amounts are >10K.

I think `assert_eq(balance, bonded);` may still be a problematic line though, as rewards disbursed could cause this assert fail. We may want a tolerance of a few credits for unbond_all.

Reply:
I think it'll be good to just claim commissions first before calling it and assert we're in the state 0u8